### ISSUE TITLE
"off by one errors" in text output

### ISSUE THREAD

**Issue by absurdist** [plantuml/plantuml#728]
once the label in a sequence diagram reaches 22 characters, the placement of the arrows is off by one space in text output 

[edit looks like there's a typo, two 0's so it's actually when it exceeds 22 characters]

```
@startuml 

participant "012345678900123456789012" as A
participant "012345678900123456789012" as B
participant "012345678900123456789012" as C
participant "012345678900123456789012" as D

A -> B : b

B -> C : c

C -> D : s

D ---> A : a

@enduml
```

produces:

```
     ┌────────────────────────┐          ┌────────────────────────┐          ┌────────────────────────┐          ┌────────────────────────┐
     │012345678900123456789012│          │012345678900123456789012│          │012345678900123456789012│          │012345678900123456789012│
     └───────────┬────────────┘          └───────────┬────────────┘          └───────────┬────────────┘          └───────────┬────────────┘
                 │                 b                 │                                   │                                   │             
                 │ ──────────────────────────────────>                                   │                                   │             
                 │                                   │                                   │                                   │             
                 │                                   │                 c                 │                                   │             
                 │                                   │ ──────────────────────────────────>                                   │             
                 │                                   │                                   │                                   │             
                 │                                   │ 
... etc.    
```

as opposed to:

```
     ┌───────────────────────┐          ┌───────────────────────┐          ┌───────────────────────┐          ┌───────────────────────┐
     │01234567890012345678901│          │01234567890012345678901│          │01234567890012345678901│          │01234567890012345678901│
     └───────────┬───────────┘          └───────────┬───────────┘          └───────────┬───────────┘          └───────────┬───────────┘
                 │                b                 │                                  │                                  │            
                 │─────────────────────────────────>│                                  │                                  │            
                 │                                  │                                  │                                  │            
                 │                                  │                c                 │                                  │            
                 │                                  │─────────────────────────────────>│                                  │            
                 │                                  │                                  │                                  │            
                 │                                  │                                  │                s                 │            
                 │                                  │   

```

https://www.plantuml.com/plantuml/uml/bSyx3a8X58JXVbOnkJt9zVqe30ysSaGB4YL4SFyUAckwRvA_6TkwVFht_I9GLJdbAgLpMJVRtVvmF9qlwv-gXTBe9be_qOQ9DWAEvaRFA--07mxg18JXg6v0f36wd2u1xBCyzEmF

**Comment by jimnelson372** [comment]
When exploring this bug, I found that it isn't just an issue when participant names are 22 or more characters, but has more to do with whether the names are even or odd lengths.   Since the timeline character can't be at a position exactly in the middle of even length participant names, it causes the message arrow to start late, or end late.   See this example:

     ┌───┐          ┌───┐          ┌────┐          ┌────┐          ┌───┐
     │010│          │020│          │0300│          │0400│          │050│
     └─┬─┘          └─┬─┘          └─┬──┘          └─┬──┘          └─┬─┘
       │   odd_odd    │              │               │               │  
       │─────────────>│              │               │               │  
       │              │              │               │               │  
       │              │  odd_even    │               │               │  
       │              │─────────────>│               │               │  
       │              │              │               │               │  
       │              │              │   even_even   │               │  
       │              │              │ ──────────────>               │  
       │              │              │               │               │  
       │              │              │               │   even_odd    │  
       │              │              │               │ ─────────────>│  
       │              │              │               │               │  
       │              │              │               │   even_odd    │  
       │              │              │               │ <─────────────│  
       │              │              │               │               │  
       │              │              │   even_even   │               │  
       │              │              │ <──────────────               │  
       │              │              │               │               │  
       │              │  odd_even    │               │               │  
       │              │<─────────────│               │               │  
       │              │              │               │               │  
       │   odd_odd    │              │               │               │  
       │<─────────────│              │               │               │  
     ┌─┴─┐          ┌─┴─┐          ┌─┴──┐          ┌─┴──┐          ┌─┴─┐
     │010│          │020│          │0300│          │0400│          │050│
     └───┘          └───┘          └────┘          └────┘          └───┘

Even length participant names position the timeline to the left of middle, while the message arrow starts to the right of "middle", so begins a character late.   For Odd lengths, the timeline is at the middle, so the arrow starts in the right place.  The arrow only ends in the wrong place (overlapping the destination timeline) when both the source and target participant names are even, since this arrow width seems to be calculated from the double values of the timeline positions rather than the integer values.   

The offset of the timelines on even named participants also impacts the Self messages:

     ┌───┐          ┌────┐            
     │010│          │0200│            
     └─┬─┘          └─┬──┘            
       ────┐          │               
           │ odd_self │               
       <───┘          │               
       │              │               
       │              │────┐          
       │              │    │ even_self
       │              │<───┘          
     ┌─┴─┐          ┌─┴──┐            
     │010│          │0200│            
     └───┘          └────┘            

Lastly, the for participants of even length, the self message label does overlap the next participant timeline when it reaches 10 characters, regardless of the even or oddness of that participant name:

     ┌────┐          ┌───┐
     │0100│          │020│
     └─┬──┘          └─┬─┘
       │────┐             
       │    │ 1234567890  
       │<───┘             
     ┌─┴──┐          ┌─┴─┐
     │0100│          │020│
     └────┘          └───┘

Here are the scripts for each of the 3 examples above:

@startuml
participant "010" as A
participant "020" as B
participant "0300" as C
participant "0400" as D
participant "050" as E

A -> B : odd_odd
B -> C : odd_even
C -> D : even_even
D -> E : even_odd
E -> D : even_odd
D -> C : even_even
C -> B : odd_even
B -> A : odd_odd

@enduml

@startuml
participant "010" as A
participant "0200" as B

A -> A : odd_self
B -> B : even_self
@enduml

@startuml
participant "0100" as A
participant "020" as B

A -> A : 1234567890
@enduml

I have made fixes that resolves these problems, but I want to test a few more conditions, and I still have one extra space on the even -> even condition that I want to double check if I can resolve that as well (it doesn't look too bad though).   (While the ascii art output may be a less important issue, I'm using this as an opportunity to learn more about the code.)

**Comment by The-Lum** [comment]
Hi all,

_[This is an Issue Review]_ 👀 
This is now fixed on [V1.2024.5](https://github.com/plantuml/plantuml/releases/tag/v1.2024.5).
- Thanks @absurdist for the report.
- Thanks PlantUML team and @jimnelson372 for your work.

Regards.
